src/QirRuntime/lib/QIR/callables.cpp

```diff
@@ -191,37 +56,207 @@ extern "C"
+        }
         callable->AddRef();
     }
 
     void quantum__rt__callable_unreference(QirCallable* callable)
     {
+        if (callable == nullptr)
+        {
+            return;
+        }
         const long ref = callable->Release();
         assert(ref >= 0);
     }
 
     QirCallable* quantum__rt__callable_create(t_CallableEntry* entries, QirTupleHeader* capture)
     {
+        assert(entries != nullptr);
         return new QirCallable(entries, capture);
     }
 
-    void quantum__rt__callable_invoke(QirCallable* clb, QirTupleHeader* args, QirTupleHeader* result)
+    void quantum__rt__callable_invoke(QirCallable* callable, QirTupleHeader* args, QirTupleHeader* result)
     {
-        clb->Invoke(args, result);
+        assert(callable != nullptr);
+        callable->Invoke(args, result);
     }
 
     QirCallable* quantum__rt__callable_copy(QirCallable* other)
     {
-        return new QirCallable(other);
+        if (other == nullptr)
+        {
+            return nullptr;
+        }
+        return new QirCallable(*other);
     }
 
-    void quantum__rt__callable_make_adjoint(QirCallable* clb)
+    void quantum__rt__callable_make_adjoint(QirCallable* callable)
     {
-        clb->ApplyFunctor(QirCallable::Adjoint);
+        assert(callable != nullptr);
+        callable->ApplyFunctor(QirCallable::Adjoint);
     }
 
-    void quantum__rt__callable_make_controlled(QirCallable* clb)
+    void quantum__rt__callable_make_controlled(QirCallable* callable)
     {
-        clb->ApplyFunctor(QirCallable::Controlled);
+        assert(callable != nullptr);
+        callable->ApplyFunctor(QirCallable::Controlled);
+    }
+}
+
+/*==============================================================================
+    Implementation of QirCallable
+==============================================================================*/
+QirCallable::~QirCallable()
+{
+    assert(refCount == 0);
+}
+
+QirCallable::QirCallable(const t_CallableEntry* ftEntries, QirTupleHeader* capture)
+    : refCount(1)
+    , capture(capture)
+    , appliedFunctor(0)
+    , controlledDepth(0)
+{
+    memcpy(this->functionTable, ftEntries, QirCallable::TableSize * sizeof(void*));
+    assert(this->functionTable[0] != nullptr); // base must be always defined
+}
+
+QirCallable::QirCallable(const QirCallable& other)
+    : refCount(1)
+    , capture(other.capture)
+    , appliedFunctor(other.appliedFunctor)
+    , controlledDepth(other.controlledDepth)
+{
+    memcpy(this->functionTable, other.functionTable, QirCallable::TableSize * sizeof(void*));
+}
+
+long QirCallable::AddRef()
+{
+    long rc = ++this->refCount;
+    assert(rc != 1); // not allowed to resurrect!
+    return rc;
+}
+
+long QirCallable::Release()
+{
+    assert(this->refCount > 0);
+
+    long rc = --this->refCount;
+    if (rc == 0)
+    {
+        delete this;
+    }
+    return rc;
+}
+
+// `src` tuple header must point to a tuple with the following structure (example for depth = 2):

```

Comment:
Suggest to add wording - we cannot verify the structure so we assume the structure is correct.